src/metabase/driver/sql/query_processor.clj

```diff
@@ -747,20 +746,20 @@
                                  (lib.metadata/field (qp.store/metadata-provider) id-or-name))
           allow-casting?       (and field-metadata
                                     (not (:qp/ignore-coercion options)))
-          database-type        (or database-type
-                                   (:database-type field-metadata))
           ;; preserve metadata attached to the original field clause, for example BigQuery temporal type information.
           identifier           (-> (apply h2x/identifier :field
                                           (concat source-table-aliases (->honeysql driver [::nfc-path source-nfc-path]) [source-alias]))
                                    (with-meta (meta field-clause)))
           identifier           (->honeysql driver identifier)
+          casted-field         (cast-field-if-needed driver field-metadata identifier)
+          database-type        (or (h2x/database-type casted-field)
```

Comment:
It looks like we might use the database type from the casted field even if allow-casting? is false. Is this desired?

Reply:
That is the point of this PR. It makes the assumption that `(:qp/ignore-coercion options)` is true, because we want to avoid casting a second time (I know for sure that this is a reason, I'm not 100% sure this is the only reason). The problem this PR wants to solve is that when a number (e.g., of type `int8`) was coerced to a `timestamp`, we use the type `timestamp` and don't revert back to `int8`.

The bug I'm fixing here resulted from using the type `int8` here. Then this field is converted to `:day`. But for postgres (because of your earlier fix), after converting to a date we cast back to the original type. That "original type" is `int8`, which breaks.